GeneratorPgn.py

```diff
@@ -0,0 +1,318 @@
+    def preprocess_batch(self, batch):
+        """
+        Applies `preprocess_fen` for a batch of data.
+        :param batch: batch to preprocess
+        :return: dictionary of preprocessed tensors
+        """
+        boards = []
+        players = []
+        castlings = []
+        en_passants = []
+        q_values = []
+        half_moves = []
+        full_moves = []
+        fens = []
+        for fen in batch:
+            data = self.preprocess_fen(fen)
+            boards.append(data['board'])
+            players.append(data['player'])
+            castlings.append(data['castling'])
+            en_passants.append(data['en_passant'])
+            half_moves.append(data['half_move'])
+            full_moves.append(data['full_move'])
+            fens.append(data['fen'])
+ 
+        return {'boards': boards,
+                'players': players,
+                'castlings': castlings,
+                'en_passants': en_passants,
+                'half_moves': half_moves,
+                'full_moves': full_moves}
+ 
+    def __iter__(self):
+        """
+        Like `Iterator.__iter__()`, but applies preprocessing to yielded data.
+        :return: yields preprocessed batches of samples
+        """
+        for batch in self.iterator:
+            #borr=chess.Board().set_fen(batch)
+            #print(borr)
+            yield self.preprocess_batch(batch),batch
+ 
+
+castling=["K","Q","k","q"]
+
+def ArrayToFen(arr):
+    """
+    converts the data returned by the olo class to fen
+    :return: list of chessboard positions in FEN notation
+    """
+    fen=[]
+    for i in range(len(arr['players'])):
+        fen.append("")
+        pieces=["P","N","B","R","Q","K","p","n","b","r","q","k"]
+        players=["w","b"]
+        for line in range(8):
+            space=0
+            for field in range(8):
+                max_value=0
+                max_value_piece=""
+                for piece in range(12):
+                    if(arr['boards'][i][piece][field][line]>max_value):
+                        max_value = arr['boards'][i][piece][field][line]
+                        max_value_piece = pieces[piece]   
+                if(max_value>0.5):
+                    if(space>0):
+                        fen[-1]=fen[-1]+str(space)
+                    fen[-1]=fen[-1]+max_value_piece
+                    space=0
+                else:
+                    space+=1
+            if(space>0):
+                fen[-1]=fen[-1]+str(space)
+            fen[-1]=fen[-1]+"/"
+        fen[-1]=fen[-1][:-1]+" "+players[int(abs(arr['players'][i]-1)/2)]
+        if (1 in arr["castlings"][i]) == False:
+            fen[-1]+=" -"
+        else:
+            fen[-1]+=" "+"".join([castling[x] for x in range(4) if arr["castlings"][i][x]==1])#en_passants
+        fen[-1]+=" -"
+        for x in range(8):  
+            if 1 in arr["en_passants"][i][0][x]:
+                fen[-1] = fen[-1][:-1]+chr(97+x)+str(1+list(arr["en_passants"][i][0][x]).index(1))

```

Comment:
this line is very hard to understand. To make it more decipherable (together with other lines in the code), the first step may be changing the names of variables from i, x to something more verbose and adding intermediate variables (e.g. instead of `+chr(97+x)+` in the middle of expression extract `chr(97+x)` to a variable and use *naming* to show what it actually means